neuralpredictors/layers/readouts/multi_readout.py

```diff
@@ -1,107 +1,147 @@
 
-        self.gamma_readout = gamma_readout  # regularisation strength
+            readout_kwargs = self.prepare_readout_kwargs(i, data_key, first_data_key, **kwargs)
 
-        for i, (k, n_neurons) in enumerate(self.neurons.items()):
             if i == 0 or clone_readout is False:
                 self.add_module(
-                    k,
-                    self._base_readout(in_shape=in_shape, outdims=n_neurons, **kwargs),
+                    data_key,
+                    self._base_readout(
+                        in_shape=in_shape_dict[data_key],
+                        outdims=n_neurons_dict[data_key],
+                        mean_activity=mean_activity,
+                        reg_weight=gamma_readout,
+                        **readout_kwargs
+                    ),
                 )
-                original_readout = k
+                original_readout = data_key
             elif i > 0 and clone_readout is True:
-                self.add_module(k, ClonedReadout(self[original_readout], **kwargs))
-
-    def initialize(self, mean_activity_dict):
-        for k, mu in mean_activity_dict.items():
-            self[k].initialize()
-            if hasattr(self[k], "bias"):
-                self[k].bias.data = mu.squeeze() - 1
-
-    def regularizer(self, readout_key):
-        return self[readout_key].feature_l1() * self.gamma_readout
-
-    @property
-    def positive(self):
-        if hasattr(self, "_positive"):
-            return self._positive
-        else:
-            return False
-
-    @positive.setter
-    def positive(self, value):
-        self._positive = value
-        for k in self:
-            self[k].positive = value
+                self.add_module(data_key, ClonedReadout(self[original_readout]))
 
+        self.initialize(mean_activity_dict)
 
-class MultiplePointPyramid2d(MultiReadout):
-    _base_readout = PointPyramid2d
+    def prepare_readout_kwargs(self, i, data_key, first_data_key, **kwargs):
+        return kwargs
 
+    def forward(self, *args, data_key=None, **kwargs):
+        if data_key is None and len(self) == 1:
+            data_key = list(self.keys())[0]
+        return self[data_key](*args, **kwargs)
 
-class MultipleGaussian3d(MultiReadout):
-    """
-    Instantiates multiple instances of Gaussian3d Readouts
-    usually used when dealing with different datasets or areas sharing the same core.
-    Args:
-        in_shape (list): shape of the input feature map [channels, width, height]
-        loaders (list):  a list of dataset objects
-        gamma_readout (float): regularisation term for the readout which is usally set to 0.0 for gaussian3d readout
-                               as it contains one dimensional weight
-
-    """
+    def initialize(self, mean_activity_dict=None):
+        for data_key, readout in self.items():
+            mean_activity = mean_activity_dict[data_key] if mean_activity_dict is not None else None
+            readout.initialize(mean_activity)
```

Comment:
Is there a reason we have these both as a method and also as input argument while initializing readout modules?

Reply:
What do you mean with "these"?

Reply:
when initializing the modules you have:

```python
self.add_module(
    data_key,
    self._base_readout(
        in_shape=in_shape_dict[data_key],
        outdims=n_neurons_dict[data_key],
        mean_activity=mean_activity,
        reg_weight=gamma_readout,
        **readout_kwargs,
        ),
    )
```

which has the `mean_activity=mean_activity` in the arguments. And every readout also has `self.initialize(mean_activity)` in their `__init__` method. Which means the `initialize()` method of `MultiReadoutBase` is redundant. right?

Reply:
You are right, it is redundant for the first initialization after instantiation (which is 99% of the cases). I would still like to argue for implementing this method also on the multireadout level, such that one can use it to re-initialize all readouts by simply calling `model.readout.initialize()`. A good point to discuss is though, whether we want to turn the `mean_activity` of the readout's `__init__` into a class attribute and then use `self.mean_activity` in the `initialize` if it's own argument `mean_activity` is None